### PR TITLE
check column limit in sort operator

### LINKED ISSUE #4135: BE crash while running sqlsmith querys when enable_pipeline_engine is true
### Steps to reproduce the behavior (Required)
1. use tpch tables with 1G data;
2. running sqlsmith

### Expected behavior (Required)
query return right result

### Real behavior (Required)
BE crash:
```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
*** Aborted at 1647327135 (unix time) try "date -d @1647327135" if you are using GNU date ***
PC: @     0x7fe22b937387 __GI_raise
*** SIGABRT (@0x3f700004224) received by PID 16932 (TID 0x7fe1bb121700) from PID 16932; stack trace: ***
    @     0x7fe22c3ec630 (unknown)
    @     0x7fe22b937387 __GI_raise
    @     0x7fe22b938a78 __GI_abort
    @          0x1867fd1 _ZN9__gnu_cxx27__verbose_terminate_handlerEv.cold
    @          0x56cd386 __cxxabiv1::__terminate()
    @          0x56cd3f1 std::terminate()
    @          0x56cd544 __cxa_throw
    @          0x1867ed8 _Znwm.cold
    @          0x2611bbc starrocks::vectorized::SortHelper::sort_on_not_null_binary_column<>()
    @          0x25e872b starrocks::vectorized::ChunksSorterFullSort::_sort_by_columns()
    @          0x25ed001 starrocks::vectorized::ChunksSorterFullSort::_sort_chunks()
    @          0x25ed147 starrocks::vectorized::ChunksSorterFullSort::done()
    @          0x258e4eb starrocks::vectorized::ChunksSorter::finish()
    @          0x2695fb1 starrocks::pipeline::PartitionSortSinkOperator::set_finishing()
    @          0x26a7d65 starrocks::pipeline::PipelineDriver::_mark_operator_finishing()
    @          0x26a7df2 starrocks::pipeline::PipelineDriver::_mark_operator_finished()
    @          0x26a8292 starrocks::pipeline::PipelineDriver::_mark_operator_cancelled()
    @          0x26a8512 starrocks::pipeline::PipelineDriver::cancel_operators()
    @          0x26a4e9a starrocks::pipeline::PipelineDriverPoller::run_internal()
    @          0x1ff24a8 starrocks::Thread::supervise_thread()
    @     0x7fe22c3e4ea5 start_thread
    @     0x7fe22b9ff8dd __clone
    @                0x0 (unknown)
```
core file:
ssh dorisdb@daily01
passwd: dorisdb@test
/home/disk1/dorisdb/qa_test_master/be
gdb ./sqlsmith_tpch_core_070b92b/starrocks_be core.16932

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
mysql> select current_version();
+------------------------+
| current_version()      |
+------------------------+
| QA_TEST_MASTER 070b92b |
+------------------------+
1 row in set (0.00 sec)
```

### LINKED ISSUE #4355: BE crash when turn on pipeline
### Steps to reproduce the behavior (Required)
1. CREATE TABLE `tpch_100g_partsupp` (
  `PS_PARTKEY` int(11) NOT NULL COMMENT "",
  `PS_SUPPKEY` int(11) NOT NULL COMMENT "",
  `PS_AVAILQTY` int(11) NOT NULL COMMENT "",
  `PS_SUPPLYCOST` decimal64(15, 2) NOT NULL COMMENT "",
  `PS_COMMENT` varchar(199) NOT NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`PS_PARTKEY`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`PS_PARTKEY`) BUCKETS 12
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
2. CREATE TABLE `tpcds_100g_store_returns` (
  `sr_item_sk` int(11) NOT NULL COMMENT "",
  `sr_ticket_number` int(11) NOT NULL COMMENT "",
  `sr_returned_date_sk` int(11) NULL COMMENT "",
  `sr_return_time_sk` int(11) NULL COMMENT "",
  `sr_customer_sk` int(11) NULL COMMENT "",
  `sr_cdemo_sk` int(11) NULL COMMENT "",
  `sr_hdemo_sk` int(11) NULL COMMENT "",
  `sr_addr_sk` int(11) NULL COMMENT "",
  `sr_store_sk` int(11) NULL COMMENT "",
  `sr_reason_sk` int(11) NULL COMMENT "",
  `sr_return_quantity` int(11) NULL COMMENT "",
  `sr_return_amt` decimal64(7, 2) NULL COMMENT "",
  `sr_return_tax` decimal64(7, 2) NULL COMMENT "",
  `sr_return_amt_inc_tax` decimal64(7, 2) NULL COMMENT "",
  `sr_fee` decimal64(7, 2) NULL COMMENT "",
  `sr_return_ship_cost` decimal64(7, 2) NULL COMMENT "",
  `sr_refunded_cash` decimal64(7, 2) NULL COMMENT "",
  `sr_reversed_charge` decimal64(7, 2) NULL COMMENT "",
  `sr_store_credit` decimal64(7, 2) NULL COMMENT "",
  `sr_net_loss` decimal64(7, 2) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`sr_item_sk`, `sr_ticket_number`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`sr_item_sk`, `sr_ticket_number`) BUCKETS 5
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
3. select   subq_0.c2 as c0,      sum(       cast(ref_1.sr_net_loss as DOUBLE)) over (partition by subq_0.c1 order by subq_0.c5) as c1,   subq_0.c0 as c2,   ref_1.sr_item_sk as c3,   subq_0.c0 as c4,   subq_0.c5 as c5,   ref_1.sr_reason_sk as c6,   subq_0.c6 as c7,   ref_1.sr_return_ship_cost as c8,   e() as c9 from   (select           ref_0.PS_PARTKEY as c0,
    ref_0.PS_SUPPKEY as c1,           ref_0.PS_SUPPLYCOST as c2,           ref_0.PS_PARTKEY as c3,           ref_0.PS_PARTKEY as c4,           ref_0.PS_COMMENT as c5,           ref_0.PS_PARTKEY as c6         from           sqlsmith.tpch_100g_partsupp as ref_0         where (true)           or (ref_0.PS_AVAILQTY = ref_0.PS_SUPPKEY)) as subq_0     left join sqlsmith.tpcds_100g_store_returns as ref_1     on (subq_0.c2 = ref_1.sr_return_amt ) where true limit 83;

### Expected behavior (Required)
Query returns correctly

### Real behavior (Required)
```
start time: Wed Mar 23 08:12:47 CST 2022
log4j:WARN No appenders could be found for logger (org.apache.hadoop.fs.FileSystem).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
*** Aborted at 1648002604 (unix time) try "date -d @1648002604" if you are using GNU date ***
PC: @     0x7f014e5a7fa4 __memcmp_sse4_1
*** SIGSEGV (@0xa1) received by PID 89473 (TID 0x7f00ec53e700) from PID 161; stack trace: ***
    @     0x7f014fa78bb5 os::Linux::chained_handler()
    @     0x7f014fa7d9b1 JVM_handle_linux_signal
    @     0x7f014fa70a48 signalHandler()
    @     0x7f014ef295f0 (unknown)
    @     0x7f014e5a7fa4 __memcmp_sse4_1
    @          0x25def2f _ZN14pdqsort_detail12pdqsort_loopIN9__gnu_cxx17__normal_iteratorIPN9starrocks10vectorized10SortHelper8SortItemINS3_5SliceEEESt6vectorIS8_SaIS8_EEEEZNS5_30sort_on_not_null_binary_columnILb0EEENS3_6StatusEPNS3_12RuntimeStateEPNS4_6ColumnEbRSA_INS4_15PermutationItemESaISK_EEmmEUlRKS8_SP_E_Lb0EEEvRKbT_ST_T0_ib.isra.0
    @          0x26106e8 starrocks::vectorized::SortHelper::sort_on_nullable_column<>()
    @          0x25f58a9 starrocks::vectorized::ChunksSorterFullSort::_sort_by_columns()
    @          0x25f7721 starrocks::vectorized::ChunksSorterFullSort::_sort_chunks()
    @          0x25f7867 starrocks::vectorized::ChunksSorterFullSort::done()
    @          0x2599930 starrocks::vectorized::ChunksSorter::finish()
    @          0x26a04d3 starrocks::pipeline::PartitionSortSinkOperator::set_finishing()
    @          0x26b20c8 starrocks::pipeline::PipelineDriver::_mark_operator_finishing()
    @          0x26b2159 starrocks::pipeline::PipelineDriver::_mark_operator_finished()
    @          0x26b26f9 starrocks::pipeline::PipelineDriver::_mark_operator_cancelled()
    @          0x26b29ea starrocks::pipeline::PipelineDriver::cancel_operators()
    @          0x26aa369 starrocks::pipeline::GlobalDriverExecutor::worker_thread()
    @          0x2007df9 starrocks::ThreadPool::dispatch_thread()
    @          0x20039b8 starrocks::Thread::supervise_thread()
    @     0x7f014ef21e65 start_thread
    @     0x7f014e53c88d __clone
    @                0x0 (unknown)
```
core file path：sqlancer01:/home/disk1/dorisdb/sqlsmith_branch_release/be/core.89473

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
branch-2.2 release c0de225

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4355 
Fixes #4135 4135

## Problem Summary(Required) ：

Check column limit in sort operator.

### CODE REVIEW

File: be/src/exec/vectorized/chunks_sorter_topn.cpp [hunk 1]
```diff
@@ -531,6 +536,11 @@ Status ChunksSorterTopn::_hybrid_sort_first_time(RuntimeState* state, Permutatio
         ++index_of_merging;
     }
 
+    if (big_chunk->reach_capacity_limit()) {
+        LOG(WARNING) << "TopN sort exceed single chunk memory limit";
+        return Status::InternalError(fmt::format("TopN sort exceed single chunk memory limit"));
```
Comment (trueeyu) [inline be/src/exec/vectorized/chunks_sorter_topn.cpp:541]: This error can be misleading, and users may easily think that the memory limit is reached.

File: be/src/exec/vectorized/chunks_sorter_topn.cpp [hunk 2]
```diff
@@ -499,6 +499,11 @@ Status ChunksSorterTopn::_hybrid_sort_common(RuntimeState* state, std::pair<Perm
         // take sort_row_number rows from permutations.second merge-sort with _merged_segment.
         _merge_sort_common(big_chunk, segments, sort_row_number, sorted_size, permutation_size, new_permutation.second);
 
+        if (big_chunk->reach_capacity_limit()) {
```
Comment (satanson) [inline be/src/exec/vectorized/chunks_sorter_topn.cpp:502]: can we use an unified macro to check capacity limit alike CHECK_MEMORY_LIMIT ?

Comment (mofeiatwork) [reply be/src/exec/vectorized/chunks_sorter_topn.cpp:502]: That's too trivial, cause every place need difference error message, and maybe got difference type of chunk.

File: be/src/exec/vectorized/chunks_sorter_topn.cpp (hunk 1 above)
Comment (mofeiatwork) [reply be/src/exec/vectorized/chunks_sorter_topn.cpp]: fixed.

**Review by trueeyu — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.2

**Pull request merged**